zeroex/order_validator.go

```diff
@@ -654,6 +685,108 @@ func (o *OrderValidator) isSupportedAssetData(assetData []byte) bool {
+	maxBinarySearchDepth := 7
+	depth := 0
+	for lo+1 < hi {
+		depth++
+		mid := (hi + lo) / 2
+		if o.isBelowMaxContentLength(signedOrders[:mid]) {
+			lo = mid
+		} else {
+			hi = mid
+		}
+		if depth == maxBinarySearchDepth {
+			break
+		}
+	}
+
+	if !o.isBelowMaxContentLength(signedOrders[:hi]) {
+		// Reject the signedOrders if it still fails at the lowest allowance
+		if hi == min {
+			// We don't ever expect this to happen since the maxOrderSizeInBytes is much smaller then the
+			// maxRequestContentLength. This error requires a single order to be larger then the
+			// maxRequestContentLength, which would only happen if someone set the maxRequestContentLength
+			// to a tiny number.
+			err := errors.New("A single signedOrder was found that was larger then the maxRequestContentLength. Increase the maxRequestContentLength")
+			log.Panic(err)
+		}
+		hi = lo
+	}
+
+	return hi
+}
+
+// isBelowMaxContentLength checks if the signedOrders once encoded into a payload is
+// below the max content length.
+func (o *OrderValidator) isBelowMaxContentLength(signedOrders []*SignedOrder) bool {
+	orders := []wrappers.OrderWithoutExchangeAddress{}
+	for _, signedOrder := range signedOrders {
+		orders = append(orders, signedOrder.ConvertToOrderWithoutExchangeAddress())
+	}
+	signatures := [][]byte{}
+	for _, signedOrder := range signedOrders {
+		signatures = append(signatures, signedOrder.Signature)
+	}
+
+	payloadSize, err := o.computePayloadSize(getOrderRelevantStatesMethodName, orders, signatures)
+	if err != nil {
+		return false
+	}
+	return payloadSize < o.maxRequestContentLength
+}
+
+func (o *OrderValidator) computePayloadSize(method string, params ...interface{}) (int64, error) {
+	data, err := o.orderValidationUtilsABI.Pack(method, params...)
+	if err != nil {
+		return int64(0), err
+	}
+	payload := map[string]interface{}{
+		"id":      2,
+		"jsonrpc": "2.0",
+		"method":  "eth_call",
+	}
+	payload["params"] = []interface{}{
+		map[string]interface{}{
+			"data": hexutil.Bytes(data),
+			"from": constants.NullAddress,
+			"to":   constants.NullAddress,
+		},
+		"latest",
+	}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return int64(0), err
+	}
+	return int64(len(payloadBytes)), nil
+}
+
+// abiHeadByteLen is the number of bytes taken up by the ABI encoded "head"
+// Source: https://solidity.readthedocs.io/en/v0.5.10/abi-spec.html#use-of-dynamic-types
+const abiHeadByteLen = int64(64)
+
+// abiArrayLenByteLen is the number of bytes taken up to describe the length of an encoded array
+// Source: https://solidity.readthedocs.io/en/v0.5.10/abi-spec.html#use-of-dynamic-types
+const abiArrayLenByteLen = int64(64)
+
+// computeNumBasicOrdersEncodable calculates the number of "basic" (e.g., non-multiAsset) 0x orders
+// that can be sent in a payload of max content size. Unlike "basic" orders, those involving the
+// MultiAssetProxy can be of arbitrary size.
+func (o *OrderValidator) computeNumBasicOrdersEncodable() int64 {
```

Comment:
It looks like all of the things we are encoding here are known ahead of time. Can we replace this method with a constant? (e.g., `const numBasicOrdersEncodable = whatever the result is`)

Reply:
The one item that is not known ahead of time is the `maxRequestContentLength` configured by the operator. We could make the rest constant but I feel it's less of a black-box to spell out all the components involved.